### ISSUE TITLE
org.redisson.RedissonRemoteService caches and runs old versions of code

### ISSUE THREAD

**Issue by LBoraz** [redisson/redisson#2498]
**Expected behavior**

When submitting a task to a Redisson executor service the task is run in the state at the time of submit().
Meaning: when i submit at task at some time X the code in the body of that task must execute exactly as it was at time X. If in the meantime the task code is modified and a new task is submitted at time Y, the code as it was submitted at time Y must be executed.

**Actual behavior**

Both code-at-time-X and code-at-time-Y are executed, randomly. This leads to unpredictable behaviour.

Actually it seems more than one version of the code is cached and one is randomly picked for execution. This was proved by adding log statements.
When restarting the app one of the versions would be executed.
It's not documented how to clear this cache.

**Steps to reproduce or test case**

BASIC:
Create a Runnable that delegates part of the execution to a spring bean. (using the bean factory obtained from the spring context)
Submit the Runnable in different versions, making sure it "prints" some different debugging output during execution, so it's possible to know which version is running.

CASE 1: in one of the versions i changed the type of data i write to a redis list from string to a json object. BOTH versions were running! That means RANDOMLY some tasks would write a json and some others a string. The service somehow caches the code to run.

CASE 2: the code caching became even more evident when the code was refactored (and one method eliminated). AFTER RESTART the executor service was expecting (for NEWLY submitted tasks) the deleted class to be present, and failed with a no-such-method-error

**Redis version**

5.0.3

**Redisson version**

3.11.6

**Redisson configuration**

**Comment by mrniko** [comment]
> When submitting a task to a Redisson executor service the task is run in the state at the time of submit().

This breaks ExecutorService's API of submit method. It doesn't gurantees that task will be in run state immediately.

> Both code-at-time-X and code-at-time-Y are executed, randomly. This leads to unpredictable behaviour.

You could chain them with RFuture listeners.

**Comment by LBoraz** [comment]
i think my bug report was not understood, looking at the answers. But i'm not using redisson anymore, so not a problem for me

**Comment by mrniko** [comment]
> If in the meantime the task code is modified and a new task is submitted at time Y

Under the same JVM used to submit task X?

**Comment by mrniko** [comment]
Here is my test. First I run this test which spins forever in seprate JVM:

    @Test
    public void runExecutor() throws InterruptedException {
        redisson.getKeys().flushall();

        Config config = createConfig();
        RedissonNodeConfig nodeConfig = new RedissonNodeConfig(config);
        nodeConfig.setExecutorServiceWorkers(Collections.singletonMap("test2", 1));
        RedissonNode node = RedissonNode.create(nodeConfig);
        node.start();

        Thread.sleep(Long.MAX_VALUE);
    }

Then I run:

    public static class SequenceTask implements Runnable, Serializable {

        public SequenceTask() {
        }

        @Override
        public void run() {
            System.out.println("value " + testMethod());
        }

        public String testMethod() {
            return "testMethod";
        }

    }

    @Test
    public void runTask() throws InterruptedException {
        RScheduledExecutorService es = redisson.getExecutorService("test2");
        for (int i = 0; i < 5; i++) {
            es.submit(new SequenceTask());
        }
    }

Then I rename testMethod to testMethod10 and run again:

    public static class SequenceTask implements Runnable, Serializable {

        public SequenceTask() {
        }

        @Override
        public void run() {
            System.out.println("value " + testMethod10());
        }

        public String testMethod10() {
            return "testMethod10";
        }

    }

    @Test
    public void runTask() throws InterruptedException {
        RScheduledExecutorService es = redisson.getExecutorService("test2");
        for (int i = 0; i < 5; i++) {
            es.submit(new SequenceTask());
        }
    }

In RedissonNode log I see follow output:

```
[main] INFO org.redisson.Version - Redisson 3.12.1.SNAPSHOT
[redisson-netty-5-8] INFO org.redisson.connection.pool.MasterPubSubConnectionPool - 1 connections initialized for 127.0.0.1/127.0.0.1:6379
[redisson-netty-5-19] INFO org.redisson.connection.pool.MasterConnectionPool - 24 connections initialized for 127.0.0.1/127.0.0.1:6379
[main] INFO org.redisson.RedissonNode - 8 map reduce worker(s) registered
[main] INFO org.redisson.RedissonNode - 1 worker(s) registered for ExecutorService with 'test' name
[main] INFO org.redisson.RedissonNode - Redisson node started!
[main] INFO org.redisson.Version - Redisson 3.12.1.SNAPSHOT
[redisson-netty-8-10] INFO org.redisson.connection.pool.MasterPubSubConnectionPool - 1 connections initialized for 127.0.0.1/127.0.0.1:6379
[redisson-netty-8-19] INFO org.redisson.connection.pool.MasterConnectionPool - 24 connections initialized for 127.0.0.1/127.0.0.1:6379
[main] INFO org.redisson.RedissonNode - 8 map reduce worker(s) registered
[main] INFO org.redisson.RedissonNode - 1 worker(s) registered for ExecutorService with 'test2' name
[main] INFO org.redisson.RedissonNode - Redisson node started!
value testMethod
value testMethod
value testMethod
value testMethod
value testMethod
value testMethod10
value testMethod10
value testMethod10
value testMethod10
value testMethod10
```

Could you point out what is wrong with my test?

**Comment by mrniko** [comment]
Two different versions of the same task run in parallel:
```
value testMethod
value testMethod10
value testMethod
value testMethod10
value testMethod
value testMethod10
value testMethod
value testMethod10
value testMethod
value testMethod10
value testMethod
value testMethod10
```
